drivers/blk/mmc/imx/usdhc.c

```diff
@@ -968,130 +1033,213 @@ void setup_blk_storage_info()
+            return;
+        }
+        driver_state.data_transfer = DRIVER_STATE_INIT;
+        if (status == DrvErrorInternal) {
+            assert("TODO; retry?");
+        }
+
+        response_status = drv_to_blk_status(status);
+        success_count = driver_state.blk_req.blk_count;
+        break;
     }
 
-    drv_status_t status;
-    switch (driver_state.executor) {
-    case ExecutorStateInit:
-        status = perform_card_identification_and_select();
-        if (status != DrvSuccess) {
+    case BLK_REQ_WRITE: {
+        drv_status_t status = usdhc_write_blocks(driver_state.blk_req.paddr,
+                                                 driver_state.blk_req.blk_number * SDDF_BLOCKS_TO_SD_BLOCKS,
+                                                 driver_state.blk_req.blk_count * SDDF_BLOCKS_TO_SD_BLOCKS);
+        if (status == DrvIrqWait) {
             return;
         }
-        setup_blk_storage_info();
+        driver_state.data_transfer = DRIVER_STATE_INIT;
+        if (status == DrvErrorInternal) {
+            assert("TODO; retry?");
+        }
 
-        driver_state.executor = ExecutorStateActive;
-        fallthrough;
+        response_status = drv_to_blk_status(status);
+        success_count = driver_state.blk_req.blk_count;
+        break;
+    }
 
-    case ExecutorStateActive:
-        handle_clients();
-        // We always stay in the Active state now.
+    default: {
+        success_count = 0;
+        response_status = BLK_RESP_ERR_INVALID_PARAM;
         break;
+    }
+    }
 
-    default:
-        assert(!"unreachable");
+    int err = blk_enqueue_resp(&blk_queue, response_status, success_count, driver_state.blk_req.id);
+    assert(!err);
+    LOG_DRIVER("Enqueued response: status=%d, success_count=%d, id=%d\n", response_status, success_count,
+               driver_state.blk_req.id);
+    microkit_notify(blk_config.virt.id);
+
+    driver_state.blk_req.inflight = false;
+
+    /* Tail-call to handle another request */
+    return handle_client(/* was_irq: */ false);
+}
+
+void do_bringup(void)
+{
+    assert(driver_status == DrvStatusBringup);
+
+    drv_status_t status = usdhc_init();
+    if (status == DrvIrqWait) {
+        return;
+    }
+    driver_state.init = DRIVER_STATE_INIT;
+    if (status != DrvSuccess) {
+        // TODO: Notify the client about the failure...?
+        clear_card_state();
+        driver_status = DrvStatusInactive;
+        LOG_DRIVER_ERR("Failed to initialise SD card\n");
+        return;
     }
+
+    setup_blk_storage_info();
+
+    /* handle any client requests that happened while initialising */
+    driver_status = DrvStatusActive;
+    handle_client(/* was_irq: */ false);
 }
 
 void notified(microkit_channel ch)
 {
+    if (driver_status == DrvStatusBringup) {
+        if (ch == device_resources.irqs[0].id) {
+            do_bringup();
+            microkit_irq_ack(ch);
+        } else {
+            /* ignore other channels */
+        }
```

Comment:
My understanding is there should be no other channels at this point, which means this should log an error/assert in this case.

Reply:
Yes this should not be possible, virtualiser will block until storage info is ready.